pkg/http/clients/base_api_client.go

```diff
@@ -0,0 +1,201 @@
+package clients
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	cerrors "github.com/contiamo/go-base/v3/pkg/errors"
+	"github.com/contiamo/go-base/v3/pkg/tokens"
+	"github.com/contiamo/go-base/v3/pkg/tracing"
+	"github.com/opentracing/opentracing-go"
+	otext "github.com/opentracing/opentracing-go/ext"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// APIError describes an error during unsuccessful HTTP request to an API
+type APIError struct {
+	// Status is the HTTP status of the API response
+	Status int
+	// Header is the set of response headers
+	Header http.Header
+	// Response is the response body
+	Response []byte
+}
+
+// Error implements the error interface
+func (e APIError) Error() string {
+	return http.StatusText(e.Status)
+}
+
+// BaseAPIClient describes all basic HTTP client operations required to work with an API
```

Comment:
```suggestion
// BaseAPIClient describes all basic HTTP client operations required to work with a JSON API
```

There are a couple spots where we are enforcing JSON req/resp, so we should be more specific here